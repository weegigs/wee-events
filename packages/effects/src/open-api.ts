import * as cc from "change-case";
import * as yaml from "yaml";
import * as z from "zod";

import { extendZodWithOpenApi, OpenAPIRegistry, OpenApiGeneratorV3 } from "@asteasolutions/zod-to-openapi";
import {
  Payload,
  Revision,
} from "@weegigs/events-core";

import { ServiceDescription } from "./service/service";

extendZodWithOpenApi(z);

export const spec = <S extends Payload>(service: ServiceDescription<any, any, S>): string => {
  const {
    entity: { name },
    title,
    version,
    description,
  } = service.info;

  const registry = new OpenAPIRegistry();
  const commands = service.commands();

  z.intersection;

  const $resource = registry.register(
    "Resource",
    z
      .object({ $self: z.string().url(), $type: z.string().min(1), $revision: Revision.schema })
      .openapi({ description: "An object with a type discriminator, location and revision" })
  );

  const resource = registry.register(
    cc.pascalCase(name),
    z.intersection($resource, service.info.entity.schema).openapi({ description: `A ${cc.pascalCase(name)} Resource` })
  );

  const notFound = registry.register(
    "NotFound",
    z
      .object({
        code: z.literal("404"),
        message: z.literal(`${name} not found`),
      })
      .openapi({
        description: `The requested ${name} was not found`,
      })
  );

  const conflict = registry.register(
    "Conflict",
    z
      .object({
        code: z.literal("409"),
        message: z.literal("conflict detected while processing request"),
      })
      .openapi({
        description: "The request could not be completed due to a conflict",
      })
  );

  const rateLimit = registry.register(
    "RateLimit",
    z
      .object({
        code: z.literal("429"),
        message: z.literal("rate limit exceeded"),
      })
      .openapi({
        description: "The request could not be completed due to a rate limit",
      })
  );

  const Id = registry.registerParameter(
    `${cc.pascalCase(name)}Id`,
    z.string().openapi({
      param: {
        name: "id",
        in: "path",
      },
      example: "1212121",
    })
  );

  registry.registerPath({
    path: `${name}/{id}`,
    method: "get",
    request: {
      params: z.object({ id: Id }),
    },
    responses: {
      200: {
        description: `A ${cc.pascalCase(name)} Resource`,
        content: {
          "application/json": { schema: resource },
        },
      },
      404: {
        description: `A Not Found Error`,
        content: {
          "application/json": { schema: notFound },
        },
      },
      429: {
        description: `A Rate Limit Error`,
        content: {
          "application/json": { schema: rateLimit },
        },
      },
    },
  });

  for (const command in commands) {
    const ref = registry.register(cc.pascalCase(command), commands[command]);

    registry.registerPath({
      method: "post",
      path: `${name}/{id}/${command}`,
      request: {
        params: z.object({ id: Id }),
        body: {
          content: {
            "application/json": {
              schema: ref,
            },
          },
        },
      },
      responses: {
        200: {
          description: `A ${cc.pascalCase(name)} Resource`,
          content: {
            "application/json": { schema: resource },
          },
        },
        404: {
          description: `A Not Found Error`,
          content: {
            "application/json": { schema: notFound },
          },
        },
        409: {
          description: `A Conflict Error`,
          content: {
            "application/json": { schema: conflict },
          },
        },
        429: {
          description: `A Rate Limit Error`,
          content: {
            "application/json": { schema: rateLimit },
          },
        },
      },
    });
  }

  const generator = new OpenApiGeneratorV3(registry.definitions);

  const serviceName = title ?? `${cc.capitalCase(name)} Service`;
  const info =
    undefined != description ? { title: serviceName, version, description } : { title: serviceName, version };
  const api = generator.generateDocument({
    openapi: "3.0.3",
    info,
  });

  return yaml.stringify(api, null, 2);
};
