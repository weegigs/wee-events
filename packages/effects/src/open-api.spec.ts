import * as api from "./open-api";
import * as receipts from "./service/sample/receipts";

describe("describe service specification", () => {
  it("should describe valid service specification", () => {
    const result = api.spec(receipts.description);
    expect(result).toMatchSnapshot();
  });
});
