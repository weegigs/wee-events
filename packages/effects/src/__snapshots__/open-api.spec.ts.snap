// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe service specification should describe valid service specification 1`] = `
"openapi: 3.0.3
info:
  title: A Sample Receipt Service
  version: 1.0.0
  description: A sample service for managing receipts
components:
  schemas:
    Resource:
      type: object
      properties:
        $self:
          type: string
          format: uri
        $type:
          type: string
          minLength: 1
        $revision:
          type: string
          minLength: 26
          maxLength: 26
      required:
        - $self
        - $type
        - $revision
      description: An object with a type discriminator, location and revision
    Receipt:
      allOf:
        - $ref: "#/components/schemas/Resource"
        - type: object
          properties:
            id:
              type: string
              minLength: 0
            total:
              type: integer
              minimum: 0
          required:
            - id
            - total
          description: A receipt
      description: A Receipt Resource
    NotFound:
      type: object
      properties:
        code:
          type: string
          enum:
            - "404"
        message:
          type: string
          enum:
            - receipt not found
      required:
        - code
        - message
      description: The requested receipt was not found
    Conflict:
      type: object
      properties:
        code:
          type: string
          enum:
            - "409"
        message:
          type: string
          enum:
            - conflict detected while processing request
      required:
        - code
        - message
      description: The request could not be completed due to a conflict
    RateLimit:
      type: object
      properties:
        code:
          type: string
          enum:
            - "429"
        message:
          type: string
          enum:
            - rate limit exceeded
      required:
        - code
        - message
      description: The request could not be completed due to a rate limit
    Add:
      type: object
      properties:
        amount:
          type: integer
          minimum: 1
      required:
        - amount
      description: Add money to the receipt
    Deduct:
      type: object
      properties:
        amount:
          type: integer
          minimum: 1
      required:
        - amount
      description: Deduct money from the receipt
    ReceiptId:
      type: string
      example: "1212121"
  parameters:
    ReceiptId:
      schema:
        $ref: "#/components/schemas/ReceiptId"
      required: true
      name: id
      in: path
paths:
  receipt/{id}:
    get:
      parameters:
        - $ref: "#/components/parameters/ReceiptId"
      responses:
        "200":
          description: A Receipt Resource
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Receipt"
        "404":
          description: A Not Found Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/NotFound"
        "429":
          description: A Rate Limit Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/RateLimit"
  receipt/{id}/add:
    post:
      parameters:
        - $ref: "#/components/parameters/ReceiptId"
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/Add"
      responses:
        "200":
          description: A Receipt Resource
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Receipt"
        "404":
          description: A Not Found Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/NotFound"
        "409":
          description: A Conflict Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Conflict"
        "429":
          description: A Rate Limit Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/RateLimit"
  receipt/{id}/deduct:
    post:
      parameters:
        - $ref: "#/components/parameters/ReceiptId"
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/Deduct"
      responses:
        "200":
          description: A Receipt Resource
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Receipt"
        "404":
          description: A Not Found Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/NotFound"
        "409":
          description: A Conflict Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Conflict"
        "429":
          description: A Rate Limit Error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/RateLimit"
"
`;
